File: measurements/article-figures.py
"""Print every figure the article quotes about the implementations, from the published data.

Run it from the repository root:

    python measurements/article-figures.py

The prose counts come from count-review-replies.ts, which pins the parser and the replies by
content. This script prints everything else: the entry positions, the unit and length medians,
the comment counts, the test totals, and the files the article names.

It scores each run with `score` from analyse-implementations.py rather than measuring anything
itself. A second definition of "named unit" or "file length" would disagree with the article by
a little, which is worse than not publishing a script at all: it reads as a check while quietly
contradicting the thing it is meant to check.

It exists because a reviewer pointed out that the two analysers between them did not produce
several published numbers, so a reader had to take those on trust.
"""
import importlib.util
import os
import re
import statistics
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
IMPLEMENTATIONS = os.path.join(HERE, "implementations")
TOOLS = ["claude", "codex", "gemini"]
ARMS = [("no rules", "control"), ("prose style", "style"), ("style + code", "code")]
PASSES = re.compile(r"(\d+) pass")


def load_scorer():
    """The analyser is the one definition of every implementation measure."""
    path = os.path.join(HERE, "analyse-implementations.py")
    spec = importlib.util.spec_from_file_location("analyse_implementations", path)
    module = importlib.util.module_from_spec(spec)
    saved = sys.argv
    sys.argv = ["analyse-implementations.py", "", ""]
    try:
        spec.loader.exec_module(module)
    finally:
        sys.argv = saved
    return module.score


score = load_scorer()


def scored(tool, arm):
    for number in range(1, 11):
        directory = os.path.join(IMPLEMENTATIONS, tool, f"{arm}-{number}")
        implementation = os.path.join(directory, "evaluate.ts")
        if os.path.isfile(implementation):
            yield f"{arm}-{number}", directory, score(implementation)


def report_positions():
    print("\nENTRY POSITION, EVERY RUN SORTED (0.00 is the top of the file)")
    for tool in TOOLS:
        for label, arm in ARMS:
            values = [row["entry"] for _, _, row in scored(tool, arm) if row["entry"] is not None]
            if values:
                shown = ", ".join(f"{value:.2f}" for value in sorted(values))
                print(f"  {tool:7s} {label:13s} median {statistics.median(values):.2f}  {shown}")


def report_positions_by_run():
    """Every Claude run named beside its entry position, sorted.

    report_positions sorts the values and drops the run names, so its output can show that a 0.21
    and a 0.11 exist without showing which runs produced them. The article names both: code-8 as
    the run nearest the boundary, and control-7 as the earliest control. code-8 is also the
    disclosed rerun, so a mistaken identity there would misreport the deviation itself.
    """
    print("\nCLAUDE ENTRY POSITION, EVERY RUN NAMED AND SORTED")
    for label, arm in ARMS:
        rows = [(row["entry"], name) for name, _, row in scored("claude", arm)
                if row["entry"] is not None]
        if not rows:
            continue
        # Three decimals, because control-7 and control-3 are both 0.11 at two, and the
        # article claims control-7 is strictly the earliest.
        shown = ", ".join(f"{name} {value:.3f}" for value, name in sorted(rows))
        print(f"  {label:13s} {shown}")


def report_shape():
    print("\nFILE SHAPE AND CORRECTNESS, MEDIAN ACROSS TEN RUNS")
    for tool in TOOLS:
        for label, arm in ARMS:
            units, longest, lengths, comments, passed, total = [], [], [], [], 0, 0
            for _, directory, row in scored(tool, arm):
                units.append(row["units"])
                longest.append(row["longest_own"])
                lengths.append(row["length"])
                comments.append(row["comments"])
                tests = os.path.join(directory, "tests.txt")
                if os.path.isfile(tests):
                    total += 1
                    found = PASSES.search(open(tests, encoding="utf-8", errors="replace").read())
                    if found and found.group(1) == "25":
                        passed += 1
            if units:
                with_comments = sum(1 for count in comments if count)
                print(f"  {tool:7s} {label:13s} units {statistics.median(units):5.1f}  "
                      f"longest unit {statistics.median(longest):5.1f}  "
                      f"file lines {statistics.median(lengths):6.1f}  "
                      f"comment lines {sum(comments):3d} across {with_comments:2d} files  "
                      f"25 of 25 in {passed} of {total}")


def report_across_arms():
    """The cross-tool figures the article quotes in prose, over all thirty files of a tool.

    Every other report here is per arm, so the aggregates the article uses when it compares one
    tool with another could not be produced from this script at all. A reader had to take them on
    trust, which is the fault this file exists to remove. The Gemini median is 64.5 rather than 65
    because its middle two values are 64 and 65.
    """
    print("\nACROSS ALL THIRTY FILES OF EACH TOOL")
    for tool in TOOLS:
        longest, comments = [], []
        for _, arm in ARMS:
            for _, _, row in scored(tool, arm):
                longest.append(row["longest_own"])
                comments.append(row["comments"])
        if not longest:
            continue
        with_comments = sum(1 for count in comments if count)
        print(f"  {tool:7s} {len(longest):2d} files  "
              f"longest own unit {summarise(longest):18s}  "
              f"comment lines {sum(comments):3d} across {with_comments:2d} files, "
              f"median {statistics.median(comments):g}")


def report_named_files():
    print("\nTHE FILES THE ARTICLE NAMES")
    # control-7 is the run the code figure shows, so its line number is printed here too.
    for tool, arm, run in [("claude", "control", 7), ("claude", "control", 8), ("claude", "code", 2)]:
        path = os.path.join(IMPLEMENTATIONS, tool, f"{arm}-{run}", "evaluate.ts")
        if not os.path.isfile(path):
            continue
        with open(path, encoding="utf-8", errors="replace") as handle:
            lines = handle.read().rstrip("\n").split("\n")
        at = next((i for i, line in enumerate(lines) if line.startswith("export function evaluate")), None)
        if at is not None:
            print(f"  {tool} {arm}-{run}: line {at + 1} of {len(lines)}")


def report_openings():
    print("\nTHE TWO OPENINGS THE ARTICLE SHOWS")
    print("  Measured by count-review-replies.ts, which holds the parser this project uses.")
    print("  Run: bun measurements/count-review-replies.ts")


def summarise(values):
    """A median with its range beside it, which the preregistration requires.

    The parentheses match the table the article already prints, so a reader comparing the two
    sees the same string rather than the same number in a different dress.
    """
    median = statistics.median(values)
    shown = f"{median:.1f}".rstrip("0").rstrip(".")
    return f"{shown} ({min(values)} to {max(values)})"


def report_preregistered():
    """The preregistered outcomes, all three arms, with the ranges the medians need.

    The five are wrapper structure as the primary outcome, then named units, longest own unit,
    comment lines and hidden tests. File length is printed below them and marked, because it is
    not one of them: a table that mixes the two silently turns an exploratory measure into a
    confirmatory one.

    The report_shape summary above rounds its medians and prints comment lines as a total, so
    neither it nor the analyser could produce the exact figures the article quotes. A median
    without its range also makes overlapping distributions look separated, which is the thing
    the range exists to prevent: named units run 9 to 15 in two of the three arms.
    """
    print("\nPREREGISTERED OUTCOMES, CLAUDE, MEDIAN (RANGE) ACROSS TEN RUNS")
    measures = [("named units", "units"), ("longest own unit", "longest_own"),
                ("comment lines", "comments")]
    columns = {}
    for label, arm in ARMS:
        rows = list(scored("claude", arm))
        if not rows:
            continue
        scores = [row for _, _, row in rows]
        column = {"wrapper files (primary)": f"{sum(1 for r in scores if r['wrapper'])} of 10"}
        for name, key in measures:
            column[name] = summarise([row[key] for row in scores])
        passed = 0
        for _, directory, _ in rows:
            tests = os.path.join(directory, "tests.txt")
            if os.path.isfile(tests):
                found = PASSES.search(open(tests, encoding="utf-8", errors="replace").read())
                if found and found.group(1) == "25":
                    passed += 1
        column["passed 25 of 25"] = f"{passed} of {len(rows)}"
        column["file lines (not preregistered)"] = summarise([row["length"] for row in scores])
        columns[label] = column
    if not columns:
        return
    labels = [label for label, _ in ARMS if label in columns]
    order = (["wrapper files (primary)"] + [name for name, _ in measures]
             + ["passed 25 of 25", "file lines (not preregistered)"])
    print(f"  {'outcome':31s}" + "".join(f"{label:>20s}" for label in labels))
    for name in order:
        print(f"  {name:31s}" + "".join(f"{columns[label][name]:>20s}" for label in labels))


if __name__ == "__main__":
    if not os.path.isdir(IMPLEMENTATIONS):
        sys.exit(f"no implementations found at {IMPLEMENTATIONS}")
    report_positions()
    report_positions_by_run()
    report_shape()
    report_preregistered()
    report_across_arms()
    report_named_files()
    report_openings()

File: measurements/article-figures.ts
/**
 * Print every figure the article quotes about the implementations, from the published data.
 *
 * Run it from the repository root:
 *
 *     bun install
 *     bun measurements/article-figures.ts
 *
 * The prose counts come from count-review-replies.ts, which pins the parser and the replies by
 * content. This script prints everything else: the entry positions, the preregistered outcomes,
 * the cross-tool aggregates, and the files the article names.
 *
 * It scores each run with `score` from score.ts rather than measuring anything itself. A second
 * definition of "named unit" would disagree with the article by a little, which is worse than not
 * publishing a script at all: it reads as a check while quietly contradicting the thing it checks.
 *
 * This replaced a pair of Python scripts that matched declarations with line-anchored regular
 * expressions. Three rounds of review each found another construct they could not see, so the
 * measurement now comes from a real parse.
 */
import { readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { score, type Score } from "./score";
import { readRerunResults } from "./rerun-results";

const MEASUREMENTS = import.meta.dir;
const IMPLEMENTATIONS = join(MEASUREMENTS, "implementations");
const TOOLS = ["claude", "codex", "gemini"] as const;
const ARMS: Array<[string, string]> = [
  ["no rules", "control"], ["prose style", "style"], ["style + code", "code"],
];

interface Run {
  name: string;
  directory: string;
  result: Score;
}

function scored(tool: string, arm: string): Run[] {
  // Every published arm holds ten runs. A short arm means a file went missing, which is reported
  // rather than divided away.
  const runs: Run[] = [];
  for (let n = 1; n <= 10; n += 1) {
    const directory = join(IMPLEMENTATIONS, tool, `${arm}-${n}`);
    const file = join(directory, "evaluate.ts");
    let text: string;
    try {
      text = readFileSync(file, "utf8");
    } catch {
      // Every one of the thirty files is counted. A run this script cannot read is a failure to
      // report, not a row to leave out of a denominator that still says ten.
      console.error(`MISSING: ${tool}/${arm}-${n} has no evaluate.ts`);
      process.exit(1);
    }
    const result = score(file, text);
    if (result.unparsed) {
      console.error(`UNPARSED: ${tool}/${arm}-${n} did not parse cleanly`);
      process.exit(1);
    }
    runs.push({ name: `${arm}-${n}`, directory, result });
  }
  return runs;
}

const median = (values: number[]): number => {
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0
    ? (sorted[middle - 1]! + sorted[middle]!) / 2
    : sorted[middle]!;
};

/** A median with its range beside it, which the preregistration requires.
 *
 * The parentheses match the table the article prints, so a reader comparing the two sees the
 * same string rather than the same number in a different dress. */
const summarise = (values: number[]): string =>
  `${median(values)} (${Math.min(...values)} to ${Math.max(...values)})`;

// The rerun's own verdicts, read once. A run absent from the manifest has not been shown to
// pass, and a manifest short of ninety runs means the battery did not finish.
const rerun = readRerunResults();
if (rerun.size !== 90) {
  console.error(`INCOMPLETE: rerun-results.txt holds ${rerun.size} runs, expected 90`);
  process.exit(1);
}

const passed = (tool: string, runs: Run[]): number =>
  runs.filter((run) => rerun.passed(`${tool}/${run.name}`)).length;

function reportPositions(): void {
  console.log("\nENTRY POSITION, EVERY RUN SORTED (0.00 is the top of the file)");
  for (const tool of TOOLS) {
    for (const [label, arm] of ARMS) {
      const values = scored(tool, arm)
        .map((run) => run.result.entry)
        .filter((value): value is number => value !== null);
      if (values.length === 0) continue;
      const shown = [...values].sort((a, b) => a - b).map((v) => v.toFixed(2)).join(", ");
      console.log(`  ${tool.padEnd(7)} ${label.padEnd(13)} median ${median(values).toFixed(2)}  ${shown}`);
    }
  }
}

function reportPositionsByRun(): void {
  // The report above sorts the values and drops the run names, so it can show that a 0.21 and a
  // 0.11 exist without showing which runs produced them. The article names both: code-8 as the
  // run nearest the boundary, and control-7 as the earliest control. Three decimals, because
  // control-7 and control-3 both read 0.11 at two and the claim is that control-7 is earliest.
  console.log("\nCLAUDE ENTRY POSITION, EVERY RUN NAMED AND SORTED");
  for (const [label, arm] of ARMS) {
    const runs = scored("claude", arm)
      .filter((run) => run.result.entry !== null)
      .sort((a, b) => (a.result.entry as number) - (b.result.entry as number));
    if (runs.length === 0) continue;
    const shown = runs.map((run) => `${run.name} ${(run.result.entry as number).toFixed(3)}`).join(", ");
    console.log(`  ${label.padEnd(13)} ${shown}`);
  }
}

function reportPreregistered(): void {
  // The five registered outcomes sit together. File length follows them and is marked, because a
  // table that mixes the two silently turns an exploratory measure into a confirmatory one.
  console.log("\nPREREGISTERED OUTCOMES, CLAUDE, MEDIAN (RANGE) ACROSS TEN RUNS");
  const columns: Array<[string, Record<string, string>]> = [];
  for (const [label, arm] of ARMS) {
    const runs = scored("claude", arm);
    if (runs.length === 0) continue;
    const results = runs.map((run) => run.result);
    columns.push([label, {
      "wrapper files (primary)": `${results.filter((r) => r.wrapper).length} of ${results.length}`,
      "named units (registered)": summarise(results.map((r) => r.registeredUnits)),
      "named units used by the wrapper rule (descriptive)": summarise(results.map((r) => r.units)),
      "longest own unit": summarise(results.map((r) => r.longest_own)),
      "comment lines": summarise(results.map((r) => r.comments)),
      "passed 25 of 25": `${passed("claude", runs)} of ${runs.length}`,
      "file lines (not preregistered)": summarise(results.map((r) => r.length)),
    }]);
  }
  if (columns.length === 0) return;
  const order = ["wrapper files (primary)", "named units (registered)",
    "named units used by the wrapper rule (descriptive)", "longest own unit", "comment lines",
    "passed 25 of 25", "file lines (not preregistered)"];
  console.log("  " + "outcome".padEnd(52) + columns.map(([label]) => label.padStart(20)).join(""));
  for (const name of order) {
    console.log("  " + name.padEnd(52) + columns.map(([, cells]) => (cells[name] ?? "").padStart(20)).join(""));
  }
}

function reportAcrossArms(): void {
  // Every other report here is per arm, so the aggregates the article uses when it compares one
  // tool with another could not be produced from this script at all.
  console.log("\nACROSS ALL THIRTY FILES OF EACH TOOL");
  for (const tool of TOOLS) {
    const results = ARMS.flatMap(([, arm]) => scored(tool, arm).map((run) => run.result));
    if (results.length === 0) continue;
    const comments = results.map((r) => r.comments);
    const withComments = comments.filter((count) => count > 0).length;
    console.log(`  ${tool.padEnd(7)} ${String(results.length).padStart(2)} files  ` +
      `longest own unit ${summarise(results.map((r) => r.longest_own)).padEnd(18)}  ` +
      `comment lines ${String(comments.reduce((a, b) => a + b, 0)).padStart(3)} across ` +
      `${String(withComments).padStart(2)} files, median ${median(comments)}`);
  }
}

function reportNamedFiles(): void {
  console.log("\nTHE FILES THE ARTICLE NAMES");
  // control-7 is the run the code figure shows, so its line number is printed here too.
  for (const [tool, arm, run] of [["claude", "control", 7], ["claude", "control", 8],
    ["claude", "code", 2]] as const) {
    const file = join(IMPLEMENTATIONS, tool, `${arm}-${run}`, "evaluate.ts");
    let lines: string[];
    try {
      lines = readFileSync(file, "utf8").replace(/\n+$/, "").split("\n");
    } catch {
      continue;
    }
    const at = lines.findIndex((line) => line.startsWith("export function evaluate"));
    if (at !== -1) console.log(`  ${tool} ${arm}-${run}: line ${at + 1} of ${lines.length}`);
  }
}

function reportOpenings(): void {
  console.log("\nTHE TWO OPENINGS THE ARTICLE SHOWS");
  console.log("  Measured by count-review-replies.ts, which holds the parser this project uses.");
  console.log("  Run: bun measurements/count-review-replies.ts");
}

try {
  readdirSync(IMPLEMENTATIONS);
} catch {
  console.error(`no implementations found at ${IMPLEMENTATIONS}`);
  process.exit(1);
}

reportPositions();
reportPositionsByRun();
reportPreregistered();
reportAcrossArms();
reportNamedFiles();
reportOpenings();
